### PR TITLE
Add the LombScargleMultiband Periodogram Class

### LINKED ISSUE #9982: Move multiband LombScargle from gatspy
Almost all of the the LS functionality has been moved to astropy from astroML and gatspy. The exception is the multiband version.

The multiple location and versions of the implementation cause some confusing amongst users, some of which I hear about, but my current response is somewhat not ideal: "always use the astropy version, as it's the same one used to be in astroML/gatspy with some extra, but if you have multiband data, well, then stick with the gatspy version"

To resolve this, I propose to do the move it here, upstream.

cc @astrofrog  as the other astropy.timeseries maintainer; and @mwcraig and @mhvk who did most of the review of the original big PR.

https://github.com/astroML/gatspy/issues/38

### PR DESCRIPTION
**Pull request by dougbrn**
### Description

This pull request is to address the need for Lomb-Scargle Periodograms that are compatible with multiband data.

This PR includes a full implementation, with test suite and documentation, of a Lomb-Scargle Periodogram class that is capable of working with multiband data. This class inherits from the existing LombScargle class, but extends it's functionality to understand bands, which are passed as an additional required parameter. Like the original port of Lomb-Scargle, this class ports the core implementation that exists in Gatspy: https://github.com/astroML/gatspy, two implementations are ready in this PR:
* Gatspy: The gatspy multiband floating mean periodogram, using matrix formalism, as outlined in VanderPlas et al. 2015 (https://arxiv.org/pdf/1502.01344.pdf)
* Fast: A conceptual port of Gatspy's fast algorithm, but leverages LombScargle directly to do band-by-band fits. 

LombScargleMultiband adheres to the conventions and API of LombScargle where possible, and the user experience is very similar. The core functions available at this time are: power, autofrequency, autopower, design_matrix, model_parameters, offset, and model.

The notable omissions from the class are:
* False alarm probabilities: LombScargle's function is not valid for models with complexity beyond a single sinusoid, common for LombScargleMultiband
* Timeseries object compatibility: Astropy Timeseries does not consider multiple bands, so we did not incorporate direct support for it, although the columns may be easily input as they are just array inputs for t, y, bands, and dy respectively.

Fixes #9982 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by dougbrn** [comment]
@mwcraig you requested a ping for reviewing this once it was in. Thanks in advance!

**Comment by dougbrn** [comment]
@pllim @mwcraig @aaryapatil I just wanted to check in on this. Is this still in the plans to be reviewed for v5.3? Let me know if there's anything I can do to assist. Thanks in advance! 🙏

**Comment by pllim** [comment]
The CI fails at multiple places. While I have pinged timeseries maintainers to review, you should rebase and fix the CI. Thanks!

**Comment by astrofrog** [comment]
I will try and review this shortly!

**Comment by astrofrog** [comment]
@dougbrn - one quick question I have before reviewing this - you mentioned that there id no compatibility with timeseries objects, but can one not represent multiband photometry with several different columns in the time series?

**Comment by dougbrn** [comment]
Hi @astrofrog , thanks for taking the time to look at this. You are right in that a user could use a column for each band. At present, lombscargle multiband expects a single flux array, with an accompanying band array that labels each measurement with a specific band label. So going from the band-per-column representation would just look like stacking the columns and adding a label column. 

However, my hesitance to support the timeseries object came from my understanding that the timeseries object has no standard way of representing multiband information. So while one user could provide column-per-band, another could provide a single flux column, with a band label column. With no standard, it wasn't clear how to best setup lombscargle multiband to handle the timeseries class. I'm open to ideas on how to get these two talking, if you have any!

**Comment by bsipocz** [comment]
fwiw, we have experimented with a multiband class and it works reasonably well out of the box, but got some pushback, or better, some non-response because it is not how some people were constructing ztf light curves (I can't recall any responses to my concerns for the labelling approach).

Needless to say, I think a multicolumn class is a much better fit for a broad usage astropy is aiming for, so ideally LS should be able to handle it.

**Comment by dougbrn** [comment]
@bsipocz If it's fair to assume from the timeseries side that multiband will only be implemented as multi-column, then I can look into getting a data loader for that format into multiband_lombscargle, I'll probably need to overload the existing from_timeseries function that is written in BasePeriodogram. Do you envision that timeseries will eventually be built out to more explicitly support this? (And if so does it make sense to wait on this at all?)

**Comment by bsipocz** [comment]
@dougbrn - It's already pretty easy to subclass TimeSeries to support multiband columns. It kind of works out of the box, unlike doing something non-standard, like adding a label in a separate column. My concerns I listed to you in the slack thread still stand.

**Comment by dougbrn** [comment]
@bsipocz Okay, this sounds good. I'll work on a commit that allows a multicolumn timeseries object as an input.

**Comment by bsipocz** [comment]
Yes, I expect it to be reasonably straightforward, basically, it could take a list of column names rather than one column name.

**Comment by dougbrn** [comment]
@bsipocz @astrofrog  Latest commit implements a from_timeseries function, with the necessary assumption of band-per-column formatting. Added some accompanying unit tests and updated the documentation. May need to push another commit or two if the CI tests fail (I've been having a finicky experience with astropy's testing of .rst files, and my local testing appears insufficient to predict the result of the CI testing...), but otherwise this should be ready for some feedback.

**Comment by dougbrn** [comment]
@astrofrog @bsipocz , this is now passing all CI tests, just in case either of you were holding off until then!

File: astropy/timeseries/periodograms/lombscargle_multiband/core.py
Comment (delucchi-cmu) [inline astropy/timeseries/periodograms/lombscargle_multiband/core.py:613]: More awkward auto-formatting.

File: astropy/timeseries/periodograms/lombscargle_multiband/implementations/main.py
Comment (delucchi-cmu) [inline astropy/timeseries/periodograms/lombscargle_multiband/implementations/main.py:4]:
```suggestion
The ``lombscarglemultiband`` function here is essentially a sophisticated switch
```

File: astropy/timeseries/periodograms/lombscargle_multiband/implementations/main.py
Comment (delucchi-cmu) [inline astropy/timeseries/periodograms/lombscargle_multiband/implementations/main.py:2]:
```suggestion
Main Lomb-Scargle Implementation with Multiband support.
```

File: astropy/timeseries/periodograms/lombscargle_multiband/implementations/main.py
Comment (delucchi-cmu) [inline astropy/timeseries/periodograms/lombscargle_multiband/implementations/main.py:13]: This const isn't referenced anywhere.

File: astropy/timeseries/periodograms/lombscargle_multiband/implementations/mle.py
Comment (delucchi-cmu) [inline astropy/timeseries/periodograms/lombscargle_multiband/implementations/mle.py:122]: Also check dimension of `y` and `dy`?

File: astropy/timeseries/periodograms/lombscargle_multiband/tests/test_lombscargle_multiband.py
Comment (delucchi-cmu) [inline astropy/timeseries/periodograms/lombscargle_multiband/tests/test_lombscargle_multiband.py:15]: I *think* if you put the pytest fixtures in a `conftest.py` file in this directory, you don't need the empty `__init__` file.

**Review by delucchi-cmu — COMMENTED** [review]
Mostly formatting/punctuation nits because I'm not familiar with the math at all.

File: astropy/timeseries/periodograms/lombscargle_multiband/implementations/mle.py
Comment (astrofrog) [inline astropy/timeseries/periodograms/lombscargle_multiband/implementations/mle.py:15]: Remove commented out code

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [inline docs/timeseries/lombscarglemb.rst:16]: Here of course it's fine to leave a reference to astroml and gatspy as in the original LombScargle page.

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [inline docs/timeseries/lombscarglemb.rst:174]: This example makes me think  this be 'uncertainty_column_name' for consistency? Or maybe it should be signal_column and uncertainty_column to make it a little more concise.

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [inline docs/timeseries/lombscarglemb.rst:183]: maybe say 'as for the LombScargle class' and link to that documentation page.

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [inline docs/timeseries/lombscarglemb.rst:185]: Is this an example of how it can be used as normal? If so, prefix with 'for example:'

**Review by astrofrog — CHANGES_REQUESTED** [review]
I apologise for taking so long to review this - this is an impressive amount of work!

I have made a number of comments below, and in addition I have two main requests:

* Would it be possible to add a test (or expand one of the existing ones) to check that ``LombScargleMultiband`` with a single band returns results consistent with ``LombScargle``?

* As this includes code copied from gatspy, we should include a copy of the gatspy license in the ``licenses`` folder. This was not a problem originally when the ``LombScargle`` class was added because it was added by @jakevdp who is the copyright holder, but we should add it now.

Once you've had a chance to reply/implement my comments, please ping me and I'll make sure we move this forward so that it can be included in the 5.3 release.

File: astropy/timeseries/periodograms/lombscargle_multiband/tests/test_lombscargle_multiband.py
Comment (astrofrog) [reply astropy/timeseries/periodograms/lombscargle_multiband/tests/test_lombscargle_multiband.py:15]: We usually always leave the ``__init__.py`` files in this package otherwise it messes with package collection by setuptools.

File: docs/timeseries/lombscarglemb.rst
Comment (dougbrn) [reply docs/timeseries/lombscarglemb.rst:140]: LombScargleMultiband treats NaN values the same way LombScargle does, which is to say that both classes don't handle them at all when passing arrays with NaN values directly through. But both check for them in the y and dy columns when ingest from a timeseries object. As for proper masked values, do you mean like from a numpy.masked.array? Under the hood it's just using np.isnan calls, which should work well with numpy's masked arrays, and from then on we treat them like numpy arrays, which I believe masked arrays naturally accommodate.

File: astropy/timeseries/periodograms/lombscargle_multiband/core.py
Comment (dougbrn) [reply astropy/timeseries/periodograms/lombscargle_multiband/core.py:15]: This is to allow for it to just be `from .implementations.mle import *` , correct? The difficulty I see with that is that we get into trouble with Flake8, (https://www.flake8rules.com/rules/F405.html), where it the imports are undefined. If we don't care about violating Flake8 that's fine, but there does seem to be an argument to be made for keeping these imports explicit.

**Comment by dougbrn** [comment]
@astrofrog , thank you for the review! It's much appreciated. I've gone through an implemented nearly all of what you've requested. I've added a test that checks for equivalent output in the single-band case for lombscarglemultiband and lombscargle, and added the gatspy license. Also, I settled on "flexible" as the replacement name for the "gatspy" method. As the method is better able to accommodate different models and input datasets thanks to its use of regularization. If you think we should come up with a different name, just let me know. I replied to a few suggestions above as well, we can talk those through and if that all looks good then I think this is in a good spot with regards to your review.

**Comment by pllim** [comment]
The commits need to be squashed. While you are at it, maybe also rebase.

Though I am not sure if this PR is too late for v5.3. It has `+2,265 −0` and it is only 3 days to feature freeze and @astrofrog requested changes but has yet to respond to the changes.

**Comment by astrofrog** [comment]
We should get this in for the feature freeze, will respond to comments this evening

**Comment by pllim** [comment]
Yes, remote data failure (timeout) is unrelated.

File: astropy/timeseries/periodograms/lombscargle_multiband/core.py
Comment (astrofrog) [reply astropy/timeseries/periodograms/lombscargle_multiband/core.py:14]: Not necessarily to enable using ``*`` but more to also restrict what *can* be imported from modules. In general we define ``__all__`` in all files - this serves as a way to be explicit about what is 'public' API.

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [reply docs/timeseries/lombscarglemb.rst:140]: As long as this is consistent with LombScargle this is fine for now - but yes I mean tables with missing values as shown here:

https://docs.astropy.org/en/stable/table/masking.html

This is the 'preferred' way of indicating missing data as opposed to using NaN values.

File: docs/changes/timeseries/14016.feature.rst
Comment (astrofrog) [inline docs/changes/timeseries/14016.feature.rst:11]: Here and below, should be ``flexible`` instead of ``gatspy``?

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [inline docs/timeseries/lombscarglemb.rst:127]: Join sentences?

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [inline docs/timeseries/lombscarglemb.rst:161]: Maybe worth printing out the table in the last cell to show what it looks like?

File: docs/timeseries/lombscarglemb.rst
Comment (astrofrog) [reply docs/timeseries/lombscarglemb.rst:140]: If you have time, could you check if masked tables work as well as NaN values? If so, I wonder if a masked table would look nicer in the docs? (but ignore for now if short on time).

File: docs/whatsnew/5.3.rst
Comment (astrofrog) [inline docs/whatsnew/5.3.rst:276]: gatspy -> flexible (also below)

File: astropy/timeseries/periodograms/lombscargle_multiband/core.py
Comment (astrofrog) [reply astropy/timeseries/periodograms/lombscargle_multiband/core.py:14]: This is to prevent users from importing say 'has_units' from this module, because they might do it and claim it was 'public API' because it worked and there were no underscores at the start of the name.

**Review by astrofrog — CHANGES_REQUESTED** [review]
This is looking good! Just a few comments, mostly small, below. The only 'big' comment is about checking whether masked tables work and if so using that in the docs instead of NaN but this is not a deal-breaker and we can postpone to a follow-up if needed.

I'll be able to review this again on Thursday and can merge then if the other comments are implemented.

**Comment by astrofrog** [comment]
@pllim - in general I wonder if we should recommend that when squashing we still keep the original commit(s) separate from the post-review ones? (rather than squashing down to one commit). Otherwise it's hard for me to check what has changed after my review, and this can be an issue in a PR this big.

File: astropy/timeseries/periodograms/lombscargle_multiband/core.py
Comment (dougbrn) [reply astropy/timeseries/periodograms/lombscargle_multiband/core.py:16]: I see, thanks for the clarification! Added an __all__ to the files that were missing it.

**Comment by dougbrn** [comment]
Sorry about the full squash, didn't realize that this wasn't what was being requested! I'll make sure to leave these subsequent commits without squashing for ease of review 👍

**Comment by pllim** [comment]
> Sorry about the full squash

It was my bad. My ask was too vague.

**Comment by eerovaher** [comment]
@dougbrn, `astropy` uses the [`pre-commit` framework](https://pre-commit.com/) which can automatically fix trailing whitespaces and more. See https://docs.astropy.org/en/stable/development/workflow/development_workflow.html#pre-commit for instructions on how to use it.

File: docs/timeseries/lombscarglemb.rst
Comment (dougbrn) [reply docs/timeseries/lombscarglemb.rst]: I just updated the from_timeseries function to support proper masked values, and updated the docs to show a masked table instead of just a table with NaN values. Also updated the test suite to use proper masking. So just to be clear, if a user creates a timeseries object and loads it into LombScargleMultiband, it recognizes proper masking and handles it accordingly. The LombScargleMultiband class itself still doesn't handle NaN values if they are included in array inputs, which is how LombScargle behaves. I would think a change to this behavior should go in a separate PR that also modifies LombScargle to keep the two consistent.

**Comment by dougbrn** [comment]
@eerovaher thanks for letting me know, have it set up now and will use it going forward.

@astrofrog, I have addressed the last few issues you flagged. Namely, adding all's to the files that were missing them, fixed the doc issues you found, and added support for proper masked values when loading in from a timeseries object. I also responded in-thread where appropriate. I rebased again, as I saw @pllim 's message on slack yesterday indicating that there was need to do so. And finally, I squashed the commits from this review into a single commit, which should keep it easy to take a look at these specific changes.

**Review by astrofrog — APPROVED** [review]
Thanks for implementing all my comments! In particular I think that in the docs it'll now be much easier for users to understand the table example (I certainly found it a lot easier to visualize myself). This is good to go, let's get it in! 😄 

Thanks again for all the work, and apologies for the delays!

**Pull request merged**
